fl4health/strategies/feddg_ga_with_adaptive_constraint.py

```diff
@@ -0,0 +1,268 @@
+        if not results:
+            return None, {}
+        # Do not aggregate if there are failures and failures are not accepted
+        if not self.accept_failures and failures:
+            return None, {}
+
+        # Convert results with packed params of model weights and training loss. The results list is modified in-place
+        # to only contain model parameters for use in the Fed-DGGA calculations and aggregation
+        train_losses_and_counts = self._unpack_weights_and_losses(results)
+
+        # Aggregate train loss
+        train_losses_aggregated = aggregate_losses(train_losses_and_counts, self.weighted_train_losses)
+        self._maybe_update_constraint_weight_param(train_losses_aggregated)
+
+        # Aggregate custom metrics if aggregation fn was provided
+        metrics_aggregated = {}
+        if self.fit_metrics_aggregation_fn:
+            fit_metrics = [(res.num_examples, res.metrics) for _, res in results]
+            metrics_aggregated = self.fit_metrics_aggregation_fn(fit_metrics)
+        elif server_round == 1:  # Only log this warning once
+            log(WARNING, "No fit_metrics_aggregation_fn provided")
+
+        self.train_metrics = {}
+        for client_proxy, fit_res in results:
+            self.train_metrics[client_proxy.cid] = fit_res.metrics
+
+        weights_aggregated = self.weight_and_aggregate_results(results)
+
+        parameters = self.parameter_packer.pack_parameters(weights_aggregated, self.loss_weight)
+        return ndarrays_to_parameters(parameters), metrics_aggregated
+
+    def _unpack_weights_and_losses(self, results: List[Tuple[ClientProxy, FitRes]]) -> List[Tuple[int, float]]:
+        """
+        This function takes results returned from a fit round from each of the participating clients and unpacks the
+        information into the appropriate objects. The parameters contained in the FitRes object are unpacked to
+        separate the model weights from the training losses. The model weights are reinserted into the parameters
+        of the FitRes objects and the losses (along with sample counts) are placed in a list and returned
+
+        NOTE: The results that are passed to this function are MODIFIED IN-PLACE
+
+        Args:
+            results (List[Tuple[ClientProxy, FitRes]]): The results produced in a fitting round by each of the clients
+                these the FitRes object contains both model weights and training losses which need to be processed.
+
+        Returns:
+            List[Tuple[int, float]]: A list of the training losses produced by client training
+        """
+        train_losses_and_counts: List[Tuple[int, float]] = []
+        for _, fit_res in results:
+            sample_count = fit_res.num_examples
+            updated_weights, train_loss = self.parameter_packer.unpack_parameters(
+                parameters_to_ndarrays(fit_res.parameters)
+            )
+            # Modify the parameters in-place to just be the model weights.
```

Comment:
Since the training losses are packed with the weights, we need to extract the losses and jam the weights back into the FitRes parameters object so that FedDG-GA can happen unabated.